Enhance and refactor Add-Type cmdlet

iSazonov opened the pull request:
## PR Summary

Fix #5515
Fix #4814
Fix #5158 

Close #5725 

- Can compile a source from parameters (TypeDefinition and MemberDefinition).
- Can compile from files.
- Can compile only to a file (without assembly loading).
- Don't recompile and don't reload if the sources have not changed.
- Implemented `-IgnoreWarnings` - ignore warnings as errors. By default the cmdlet consider warnings as errors.
- Added VisualBasic support.
- Added new `-CompilerOptions` parameter allow set Roslyn command line parameters including:
    - Parser options.
    - Compile options.
    - Emit options.

**ATTENTION:** The `CompilerOptions` can be specified along with other options like `-OutputAssembly`, `-Language` and `-IgnoreWarnings`. The explicit setting parameters will take precedence over the same settings specified in `-CompileOptions`.

See docs about the compiler options:
https://github.com/dotnet/roslyn/blob/master/docs/compilers/CSharp/CommandLine.md
https://github.com/dotnet/roslyn/blob/master/docs/compilers/Visual%20Basic/CommandLine.md

**ATTENTION:** `-OutputType` default is `Library`. If `-OutputType` is absent the `-OutputType` default overlaps a value in `CompileOptions`. In other words output type ("target" ot "t" in command line) is always ignored in `CompileOptions`. We have to use `-OutputType` to set an output type.

@TravisEz13 @PaulHigin Please audit security of the `-ExtendedOptions` parameter.

## PR Checklist

Note: Please mark anything not applicable to this PR `NA`.

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [x] Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] User facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [ ] Issue filed - Issue link:
- [ ] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
    - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.

iSazonov committed:
Enhance and refactor Add-Type cmdlet

iSazonov committed:
Revert ReferencedAssemblies to silently ignore null value

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -111,75 +78,65 @@ public sealed class AddTypeCommand : PSCmdlet
-
-        internal string typeNamespace = "Microsoft.PowerShell.Commands.AddType.AutoGeneratedTypes";
+        [Alias("NS")]
+        public String Namespace  = "Microsoft.PowerShell.Commands.AddType.AutoGeneratedTypes";
```
Fixed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -111,75 +78,65 @@ public sealed class AddTypeCommand : PSCmdlet
         /// </summary>
         [Parameter(ParameterSetName = "FromMember")]
+        [Parameter(ParameterSetName = "FromSource")]
+        [ValidateNotNullOrEmpty()]
```
The default is Utils.EmptyArray<string>() - with null argument we get an exception in a code.

iSazonov replied:
Good catch!
Fixed.

iSazonov committed:
Convert Namespace to property

iSazonov committed:
Replace [ValidateNotNullOrEmpty()] with [ValidateNotNull()] in UsingNamespace

iSazonov committed:
Load assembly from MemoryStream

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Add-Type.Tests.ps1:
```diff
@@ -26,41 +26,212 @@ Describe "Add-Type" -Tags "CI" {
+    ) {
+        param($type1, $type2, $file1, $file2, $sourceLanguage)
+
+        { $type1 = Invoke-Expression -Command $type1 } | Should Throw
```
this could be done also as:
```
$type1 -as [type] | Should BeNullOrEmpty
```
etc, yes?

iSazonov replied:
Good catch! Thanks!

Fixed.

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Add-Type.Tests.ps1:
```diff
@@ -26,41 +26,212 @@ Describe "Add-Type" -Tags "CI" {
+        { Invoke-Expression -Command $cmdlet } | Should Throw
+
+        $testModule = Import-Module -Name $outFile -PassThru
+        Invoke-Expression -Command $cmdlet | Should BeExactly $guid
```
why is this not just
`& $cmdlet`

iSazonov replied:
Thanks! Fixed.

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Add-Type.Tests.ps1:
```diff
@@ -26,41 +26,212 @@ Describe "Add-Type" -Tags "CI" {
+        # Catch non-termination information error for ExtendedOptions.
+        { Add-Type -ExtendedOptions "/platform:anycpuERROR" -Language CSharp -MemberDefinition "public static string TestString() { return ""}" -Name "TestType1" -Namespace "TestNS" -ErrorAction Stop } | ShouldBeErrorId "SOURCE_CODE_ERROR,Microsoft.PowerShell.Commands.AddTypeCommand"
+        { Add-Type -ExtendedOptions "/platform:anycpuERROR" -Language VisualBasic -MemberDefinition "Public Shared Function TestString() As String `n Return `"`" `n End Function" -Name "TestType1" -Namespace "TestNS" -ErrorAction Stop } | ShouldBeErrorId "SOURCE_CODE_ERROR,Microsoft.PowerShell.Commands.AddTypeCommand"
+    }
```
can we use the new pester builtin for checking errorid?
```
{ Add-Type -ExtendedOptions "/platform:anycpuERROR" -Language CSharp -MemberDefinition "public static string TestString() { return ""}" -Name "TestType1" -Namespace "TestNS" -ErrorAction Stop } | Should -throw -errorid "SOURCE_CODE_ERROR,Microsoft.PowerShell.Commands.AddTypeCommand"
```

iSazonov replied:
If you approve we should change the pattern in our docs https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/WritingPesterTests.md and then fix all tests.

iSazonov replied:
Currently we have a tracking Issue to fix this over all tests.

iSazonov committed:
Fix test patterns

vors commented on src/Microsoft.PowerShell.Commands.Utility/Microsoft.PowerShell.Commands.Utility.csproj:
```diff
@@ -78,6 +78,7 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="2.6.1" />
+    <PackageReference Include="Microsoft.CodeAnalysis.VisualBasic" Version="2.6.1" />
```
I remember we explicitly decided not to support VisualBasic in pwsh implementation because it's not used and increases the size of the package.

iSazonov replied:
In previous cleanup PR we got PowerShell Committee conclusion about Version type - nothing has been said that VisualBasic  is not necessary 😕 Also we have request about adding F#.. And JS (it seems impossible).

iSazonov replied:
I could not make a common code for C# and VB - each language requires its own code path. So I have no idea how we can add languages as a package.

vors commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -568,19 +560,15 @@ private string GetUsingTemplate(Language language)
             {
                 case Language.CSharp:
                     return
-                        "using System;\n" +
```
This is breaking change compare to ps 5.1

iSazonov replied:
I'd revert this but the _whole_ cmdlet is a breaking change. So it makes no sense to revert.

iSazonov replied:
Fixed.

vors commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1090,36 +1252,166 @@ private void CompileSourceToAssembly(string source)
+
+                var lineSpan = location.GetLineSpan(); // FileLinePositionSpan type.
+                var errorLineNumber = lineSpan.StartLinePosition.Line;
+                var errorPosition = lineSpan.StartLinePosition.Character;
```
errorLineNumber and errorPosition should be part of the error message

iSazonov replied:
We get this implicitly in next line.

vors replied:
Ah in the `diagnisticMessage`, I see. Probably worth to add comments here, for example to the docstring of this method to show generated error text for the giving input.

iSazonov replied:
The API haven't docs. The message format can be changed in Roslyn repo. So I believe we should add short custom comment.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -109,75 +77,65 @@ public sealed class AddTypeCommand : PSCmdlet
-
-        internal string typeNamespace = "Microsoft.PowerShell.Commands.AddType.AutoGeneratedTypes";
+        [Alias("NS")]
+        public String Namespace { get; set; }  = "Microsoft.PowerShell.Commands.AddType.AutoGeneratedTypes";
```
It was in original Windows PowerShell code. I don't know whether we block this. Seems this allows bold code injections.

PaulHigin replied:
Ok, I think we should keep it then.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -317,39 +275,30 @@ private void ProcessPaths(List<string> resolvedPaths)
 
-        internal String[] assemblyNames;
-        internal bool loadAssembly = false;
+        private bool loadAssembly = false;
```
It is in Line 244.

PaulHigin replied:
I see it now.  Thanks.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -459,51 +408,94 @@ public string OutputAssembly
+
         /// <summary>
         /// Flag to ignore warnings during compilation.
+        /// Not implemented.
```
Will remove the old comment.

iSazonov replied:
Fixed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -459,51 +408,94 @@ public string OutputAssembly
+        [Parameter(ParameterSetName = "FromPath")]
+        [Parameter(ParameterSetName = "FromLiteralPath")]
+        [Alias("IW", "IgnoreWarnings")]
+        public SwitchParameter WarnAsError { get; set; }
```
Good catch! They is reverse.
The idea was to get a more understandable and familiar name.
Need more feedback.
Maybe leave IgnoreWarnings for backward compatibility and use new WarnAsError?

PaulHigin replied:
>> Maybe leave IgnoreWarnings for backward compatibility and use new WarnAsError?
Yes, I like this better.  But maybe "WarningsAsErrors"?

iSazonov replied:
Another thought - what about rename `WarnAsError` to `NoWarningsAsErrors`? So we keep one parameter and backward compatible behavior.

/cc @daxian-dbw

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -714,75 +614,91 @@ protected override void EndProcessing()
+        private bool InMemory { get { return String.IsNullOrEmpty(outputAssembly); } }
+
+        // These dictionaries prevent reloading already loaded and unchanged code.
+        private static Dictionary<string, int> s_sourceTypesCache = new Dictionary<string, int>();
```
I think it makes no sense until we have the opportunity to unload the compiled assemblies.

PaulHigin replied:
Ok, makes sense.

iSazonov replied:
I added TODO comment to review this after migrating to .Net Core 2.1.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -459,51 +408,94 @@ public string OutputAssembly
         public SwitchParameter PassThru { get; set; }
 
+        /// <summary>
+        /// Flag to don't load compiled assembly.
```
Will change.

iSazonov replied:
Fixed.

iSazonov committed:
Fix comments

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +870,379 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+                PathType.GetRandomFileName(),
+                syntaxTrees: _syntaxTrees,
+                references: references,
+                options: _csharpCompilationOptions);
```
`_csharpCompilationOptions` can also be made a local variable in `CSharpSourceCodeProcessing()`, by changing this method to take a parameter of `CompilationOptions compilationOptions`.

iSazonov replied:
Fixed

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +870,379 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+                    _csharpCompilationOptions = arguments.CompilationOptions.WithGeneralDiagnosticOption(s_generalDiagnosticOption);
+                }
+
+                _emitOptions = arguments.EmitOptions;
```
Same here, `_emitOptions` can be made a local variable, by adding another parameter to `CSharpCompileToAssembly()` as `EmitOption emitOption` and also pass the emit option to `DoEmitAndLoadAssemply(Compilation)`.

iSazonov replied:
Fixed

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +870,379 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+                    using (var sourceFile = new FileStream(file, FileMode.Open))
+                    {
+                        var sourceText = SourceText.From(sourceFile);
+                        _syntaxTrees.Add(CSharpSyntaxTree.ParseText(sourceText, _csharpParseOptions, file));
```
Will this handle the case that the code in those files references to symbols defined in other files?

iSazonov replied:
My understanding is yes - it is the same parser/compiler session. But I don't test this.

daxian-dbw replied:
Can you please verify it? `ParseText` is a static method, so it's hard to say for sure that one signle parser session is used for all those files.

iSazonov replied:
Done. Works well.
```powershell
Add-Type -Path C:\tmp\1.cs,c:\tmp\2.cs
[Test.AddType.CSharpTest2]::Add2(1,2)
3
```

1.cs
```cs
namespace Test.AddType
        {
            public class CSharpTest1
            {
                public static int Add1(int a, int b)
                {
                    return (a + b);
                }
            }
        }
```

2.cs
```cs
namespace Test.AddType
        {
            public class CSharpTest2
            {
                public static int Add2(int a, int b)
                {
                    return CSharpTest1.Add1(a, b);
                }
            }
        }
```

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +870,379 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+            else
+            {
+                Diagnostics.Assert(false, "Invalid parameter set: {0}", this.ParameterSetName);
+            }
```
nit: By looking at this assertion here, I feel maybe arrange this `if {} else if {} else if {} else {}` in a switch of parameter name would be cleaner.

iSazonov replied:
Done.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +870,379 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+            {
+                // if the source code was already compiled and loaded and not changed
+                // we get the assembly from the cache.
+                if (isSourceCodeUpdated(out Assembly assembly))
```
I think it would be better if you make `_syntaxTree` also a local variable. It can be passed to `IsSourceCodeUpdated` and `CSharpCompileToAssembly`.
I think it's better to make it clear what a method consumes and produces in its signature, and having the arguments like `compilationOptions, emitOptions, syntaxTree` passed into methods helps it.

iSazonov replied:
I am waiting https://github.com/PowerShell/PowerShell/pull/6141#discussion_r171310625

iSazonov replied:
Fixed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +870,379 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+                    _visualbasicCompilationOptions = arguments.CompilationOptions.WithGeneralDiagnosticOption(s_generalDiagnosticOption);
+                }
+
+                _emitOptions = arguments.EmitOptions;
```
Similar comments as in `CSharpSourceCodeProcessing`.

iSazonov replied:
Fixed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +870,379 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+                // we get the assembly from the cache.
+                if (isSourceCodeUpdated(out Assembly assembly))
+                {
+                    CSharpCompileToAssembly();
```
Do we honor `-PassThru` in this case?

iSazonov replied:
Yes, see below in `DoEmitAndLoadAssemply`

daxian-dbw reviewed (changes requested):
Haven't finished all but want to share the feedback early. Will continue review from `CheckDuplicateTypes`.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -62,42 +66,6 @@ public enum OutputAssemblyType
-    /// <summary>
-    /// Compile error or warning.
-    /// </summary>
-    public class AddTypeCompilerError
```
Done.

daxian-dbw replied:
#Closed

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -109,75 +77,65 @@ public sealed class AddTypeCommand : PSCmdlet
         /// Any using statements required by the auto-generated type.
         /// </summary>
         [Parameter(ParameterSetName = "FromMember")]
+        [Parameter(ParameterSetName = "FromSource")]
```
It allows compile a source code on PowerShell Core and Windows PowerShell with different set of usings. Sometimes .Net Core and .Net Framework is different in that.

daxian-dbw replied:
But the `Add-Type` in Windows PowerShell doesn't support this parameter in `FromSource` set, and thus it won't bring us this benefit. I would rather have `UsingNamespace` only belong to `FromMember` set, just like today.

iSazonov replied:
I believe the scenario can be important if someone has a portable code template and have to compile in Windows PowerShell and PowerShell Core with different usings. If the scenario is important we should enhance Windows PowerShell too. If no I'll remove this.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -317,39 +275,30 @@ private void ProcessPaths(List<string> resolvedPaths)
         /// </summary>
         [Parameter(ParameterSetName = "FromSource")]
         [Parameter(ParameterSetName = "FromMember")]
+        [Parameter(ParameterSetName = "FromPath")]
```
Yes, will do. (I wanted reduce size of the PR)

iSazonov replied:
Done.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -459,51 +408,94 @@ public string OutputAssembly
+        [Parameter(ParameterSetName = "FromMember")]
+        [Parameter(ParameterSetName = "FromPath")]
+        [Parameter(ParameterSetName = "FromLiteralPath")]
+        public SwitchParameter CompileOnly { get; set; }
```
We should always load assembly by default otherwise it is a breaking change.

daxian-dbw replied:
`Add-Type -TypeDefinition $code -OutputAssembly f:\test.dll` doesn't load the assembly today, so why would it be a breaking change?

I tried the following in PSCore:
```powershell
[System.AppDomain]::CurrentDomain.GetAssemblies() | Out-File F:\old.txt
Add-Type -TypeDefinition $code -OutputAssembly F:\test.dll
[System.AppDomain]::CurrentDomain.GetAssemblies() | Out-File F:\new.txt
```
One assembly is loaded when calling `Add-Type`, and it's `System.ValueTuple.dll`. I don't see the `test.dll` being loaded.

iSazonov replied:
Fixed

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -459,51 +408,94 @@ public string OutputAssembly
+        [Parameter(ParameterSetName = "FromPath")]
+        [Parameter(ParameterSetName = "FromLiteralPath")]
+        [Alias("IW", "IgnoreWarnings")]
+        public SwitchParameter WarnAsError { get; set; }
```
Yes, @PaulHigin catched this above - the both is reverse.
Suggestion is to keep `IgnoreWarnings` for backward compatibility and add new `WarningsAsErrors` as  a more understandable and familiar name.

iSazonov replied:
Please see my new suggestion above.

daxian-dbw replied:
In Windows PowerShell, `Add-Type` treats warning as an error by default. It ignores compilation warnings only if `-IgnoreWarnings` is specified. See the example below:
```powershell
$code = @'
using System;

[Obsolete("Obsolete test")]
public class Foo {
public string Name;
}

public class Bar {
public Foo foo;
}
'@

PS:13> Add-Type -TypeDefinition $code
Add-Type : c:\Users\dongbow\AppData\Local\Temp\ojdqnm22.0.cs(8) : Warning as Error: 'Foo' is obsolete: 'Obsolete test'
c:\Users\dongbow\AppData\Local\Temp\ojdqnm22.0.cs(7) : public class Bar {
c:\Users\dongbow\AppData\Local\Temp\ojdqnm22.0.cs(8) : >>> public Foo foo;
c:\Users\dongbow\AppData\Local\Temp\ojdqnm22.0.cs(9) : }
At line:1 char:1
+ Add-Type -TypeDefinition $code
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (Microsoft.Power...peCompilerError:AddTypeCompilerError) [Add-Type], Except
   ion
    + FullyQualifiedErrorId : SOURCE_CODE_ERROR,Microsoft.PowerShell.Commands.AddTypeCommand

Add-Type : Cannot add type. Compilation errors occurred.
At line:1 char:1
+ Add-Type -TypeDefinition $code
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Add-Type], InvalidOperationException
    + FullyQualifiedErrorId : COMPILER_ERRORS,Microsoft.PowerShell.Commands.AddTypeCommand

PS:14> Add-Type -TypeDefinition $code -IgnoreWarnings
# successful
```
The same behavior should be retained -- treat warnings as errors by default but allowing `-IgnoreWarnings` to suppress.

iSazonov replied:
Roslyn has [opposite default](https://github.com/dotnet/roslyn/blob/944ae114140ba77cbd4be370bf13a7f758f740b3/src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs#L744). 😕 Should we keep CodeDom default? I'd prefer use new parameter and new default and keep old parameter and default for backward compatibility only.

daxian-dbw replied:
We should honor the existing behavior in Windows PowerShell, as long as that behavior is reasonable. This is important for script portability.

iSazonov replied:
Fixed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -459,51 +408,94 @@ public string OutputAssembly
+        /// Parser options:
+        ///     langversion:string - language version from:
+        ///                                 [enum]::GetNames([Microsoft.CodeAnalysis.CSharp.LanguageVersion])
+        ///                                 [enum]::GetNames([Microsoft.CodeAnalysis.CSharp.LanguageVersion])
```
It's typo. Will remove.

iSazonov replied:
Ah, second is for VisualBasic. Fixed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -459,51 +408,94 @@ public string OutputAssembly
+        [Parameter(ParameterSetName = "FromPath")]
+        [Parameter(ParameterSetName = "FromLiteralPath")]
+        [Alias("EP")]
+        public string[] ExtendedOptions { get; set; }
```
The name is from Roslyn. It is used for parser, compile and emit. It is Roslyn command line arguments.

iSazonov replied:
Alias is removed.

daxian-dbw replied:
https://github.com/dotnet/roslyn/blob/master/docs/compilers/CSharp/CommandLine.md
https://github.com/dotnet/roslyn/blob/master/docs/compilers/Visual%20Basic/CommandLine.md

As shown in the above pages, these options are called `Visual C# Compiler Options` and `Visual Basic Compiler Options`, so I think `CompilerOptions` would be a better name.

iSazonov replied:
Clear. Renamed.

iSazonov replied:
Fixed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -714,75 +614,91 @@ protected override void EndProcessing()
+        protected override void EndProcessing()
+        {
+            // The CompileOnly parameter requires that the OutputAssembly parameter be specified.
+            if (CompileOnly && String.IsNullOrEmpty(outputAssembly))
```
Yes, we should load assembly by default as in Windows PowerShell.

daxian-dbw replied:
See https://github.com/PowerShell/PowerShell/pull/6141#discussion_r173548138
I don't think `Add-Type -TypeDefinition $code -OutputAssembly f:\test.dll` loads the assembly in Windows PowerShell.

iSazonov replied:
Sorry, I already forget my thoughts - `Add-Type -TypeDefinition $code -OutputAssembly f:\test.dll` in Windows PowerShell loads the dll only if `-PassTru` present. It is hidden logic and the idea is add new parameter for better UX.

iSazonov replied:
Fixed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -714,75 +614,91 @@ protected override void EndProcessing()
+
+        private static readonly string s_defaultSdkDirectory = System.Runtime.InteropServices.RuntimeEnvironment.GetRuntimeDirectory();
+        private static ReportDiagnostic s_generalDiagnosticOption = ReportDiagnostic.Error;
+        private static string[] s_writeInformationTags = new string[] { "PSHOST" };
```
Idea is to write on host not to information stream.

iSazonov replied:
My thought is to write Roslyn info messages to the host by default. Otherwise the user may not guess that they are.

daxian-dbw replied:
OK, understood. #closed

iSazonov committed:
Add TODO comment about cache size and unloading assemblies

iSazonov committed:
Add parameter set name constants

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -459,51 +408,94 @@ public string OutputAssembly
+        [Parameter(ParameterSetName = "FromPath")]
+        [Parameter(ParameterSetName = "FromLiteralPath")]
+        [Parameter(ParameterSetName = "FromAssemblyName")]
         public SwitchParameter PassThru { get; set; }
```
Currently it is not in "FromAssemblyName".
I think it makes sense to add?

daxian-dbw replied:
`Add-Type -AssemblyName F:\pscore\Microsoft.PowerShell.Commands.Utility.dll -PassThru` returns all available types from the DLL, so I think we should keep the same.

iSazonov replied:
This works.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -459,51 +408,94 @@ public string OutputAssembly
+
         /// <summary>
         /// Flag to ignore warnings during compilation.
+        /// Not implemented.
```
The line is removed.

iSazonov committed:
Fix typo in comment

iSazonov committed:
Remove alias for ExtendedOptions

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -714,75 +614,91 @@ protected override void EndProcessing()
 
-            // Generate an error if they've specified an output
-            // assembly type without an output assembly
-            if (String.IsNullOrEmpty(outputAssembly) && outputTypeSpecified)
```
`OutputType` is translated in CompilationOptions. Since we have ExtendedOptions parameter we can catch tons conflicts - we'll get runtime errors. I guess we won't be able to identify all conflicts- it is better to delegate this to Roslyn.

daxian-dbw replied:
Then why do you check this case: `Add-Type $code -CompileOnly` without `-OutputAssemby`?
They look to me are the same situation.

iSazonov replied:
Fixed.

iSazonov committed:
Use Utils.DefaultPowerShellAppBase  for SDK home

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -714,75 +614,91 @@ protected override void EndProcessing()
+        private static Dictionary<string, int> s_sourceTypesCache = new Dictionary<string, int>();
+        private static Dictionary<int, Assembly> s_sourceAssemblyCache = new Dictionary<int, Assembly>();
+
+        private static readonly string s_defaultSdkDirectory = System.Runtime.InteropServices.RuntimeEnvironment.GetRuntimeDirectory();
```
Fixed.

iSazonov committed:
Replace s_generalDiagnosticOption with defaultDiagnosticOption

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -714,75 +614,91 @@ protected override void EndProcessing()
+        private static Dictionary<int, Assembly> s_sourceAssemblyCache = new Dictionary<int, Assembly>();
+
+        private static readonly string s_defaultSdkDirectory = System.Runtime.InteropServices.RuntimeEnvironment.GetRuntimeDirectory();
+        private static ReportDiagnostic s_generalDiagnosticOption = ReportDiagnostic.Error;
```
Agree. Fixed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -803,77 +719,21 @@ private void LoadAssemblies(IEnumerable<string> assemblies)
         {
             // netcoreapp2.0 currently comes with 137 reference assemblies (maybe more in future), so we use a capacity of '150'.
             var defaultRefAssemblies = new List<PortableExecutableReference>(150);
+            Diagnostics.Assert(defaultRefAssemblies.Count < 150, "Size of defaultRefAssemblies is small.");
```
Now we reach this threshold. We can exceed it after the transition to .Net Core 2.1. That's why I made this reminder.

daxian-dbw replied:
There are two problems here:
1. The assert at this location doesn't do anything useful because when the list is initially created, its count will be 0.
2. I don't know how many people uses debug mode binaries. It could happen that this assert fails without us knowing. This happened before to an assert in `NativeComamndParameterBinder`.

iSazonov replied:
Clear. Removed.

iSazonov committed:
Use this.SessionState.Path.CurrentLocation.Path as baseDirectory

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +870,379 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+            }
+        }
+
+        private static CSharpCommandLineArguments CSharpDefaultParse(IEnumerable<string> args, string baseDirectory = "./", string sdkDirectory = null, string additionalReferenceDirectories = null)
```
Good catch! We can not use this in static method but we can use this as named argument at a call point.
Fixed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +870,379 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+            return CSharpCommandLineParser.Default.Parse(args, baseDirectory, sdkDirectory, additionalReferenceDirectories);
+        }
+
+        private static VisualBasicCommandLineArguments VisualBasicDefaultParse(IEnumerable<string> args, string baseDirectory = "./", string sdkDirectory = null, string additionalReferenceDirectories = null)
```
Fixed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +870,379 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+                HandleCompilerErrors(arguments.Errors);
+
+                _csharpParseOptions = arguments.ParseOptions;
+                _csharpCompilationOptions = arguments.CompilationOptions.WithOutputKind(OutputAssemblyTypeToOutputKind(OutputType));
```
Yes, it was my headache. I suggest that (1) parameters have high priority, (2) delegate conflict detections to Roslyn as I said above we can not predict all.

daxian-dbw replied:
In Windows PowerShell, `Add-Type` has `-CompilerParameters` parameter which is a group of options similar to the extended options here. The approach used there is to throw error when the conflict happens:
```
            // Generate error if user has specified CompilerParmeters and
            // (referencedAssemblies | ignoreWarnings | outputAssembly | outputType)
```
We can consider doing the same.
Or maybe the `extended options` can be put in a parameter set that is for advanced user scenarios. Other parameters that may possibly cause contradiction are put in other exclusive parameter sets.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +870,379 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+
+                HandleCompilerErrors(arguments.Errors);
+
+                _csharpParseOptions = arguments.ParseOptions;
```
I tried to make the code simpler and more _general_. I could not combine the code for C# and VB but left these fields to make it easier to add enhancements.

I'll wait until you finish the review pass - maybe you'll have ideas on how to make the code structure more unified. Then we'll go back to this place.

iSazonov replied:
Fixed.

iSazonov committed:
Use switch for ParameterSets

iSazonov committed:
Add comment about diagnostic message

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1090,36 +1259,168 @@ private void CompileSourceToAssembly(string source)
+
+                var lineSpan = location.GetLineSpan(); // FileLinePositionSpan type.
+                var errorLineNumber = lineSpan.StartLinePosition.Line;
+                var errorPosition = lineSpan.StartLinePosition.Character;
```
`errorPosition` is not used anywhere. Do you intend to use it to replace `lineSpan.StartLinePosition.Character` at line 1397?

iSazonov replied:
Yes, thanks - it's more readable.
Done.

daxian-dbw replied:
I don't see the change that addresses this comment in your new commit.

iSazonov replied:
Sorry, lost commit.
Fixed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1090,36 +1259,168 @@ private void CompileSourceToAssembly(string source)
-                    Column = d.Location.GetMappedLineSpan().StartLinePosition.Character + 1, // Convert 0-based to 1-based
-                    ErrorNumber = d.Id
-                }).ToArray();
+                    if (errorLineString[i] != ' ' && errorLineString[i] != '\t')
```
Maybe change to `if (!char.IsWhiteSpace(errorLineString[i]))`? It's more readable.

iSazonov replied:
I thought about it but there's a lot of other characters
https://msdn.microsoft.com/en-us/library/t809ektx%28v=vs.110%29.aspx
Is it ok?

daxian-dbw replied:
As long as they are whitespace characters, I don't see a difference. Do you?

iSazonov replied:
I fixed this but ...
This come from original code and I'm afraid the code isn't exactly correct because we're replacing one tab with a single space. Perhaps it was necessary for CodeDom. I hope that Roslyn don't use tabs in error messages. Although these whitespaces can come from the code that it compiles. Therefore we have to clean the error message too. Thoughts?

iSazonov replied:
Fixed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1090,36 +1259,168 @@ private void CompileSourceToAssembly(string source)
+                        // We copy white chars from the source string.
+                        sb.Append(errorLineString, 0, i);
+                        // then pad up to the error position.
+                        sb.Append(' ', Math.Max(0, lineSpan.StartLinePosition.Character - i));
```
> Math.Max(0, lineSpan.StartLinePosition.Character - i)

Is the `Math.Max` call really needed? It seems to me that `errorPosition` would be `-ge i`.

iSazonov replied:
When I researched how to make the error report I was looking for code examples. Now I do not remember where I saw this pattern.
I think this is because the lineSpan and the error position are completely different code paths - so we should protect.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -449,61 +398,102 @@ public string OutputAssembly
+        ///     nowarn                   - disable all warnings
+        ///     nowarn:strings           - disable a list of individual warnings
+        ///     usings:strings           - ';'-delimited usings for CSharp
+        ///     imports:strings          - ';'-delimited imports for CSharp
```
For CSharp or VB?

iSazonov replied:
VisualBasic
Fixed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -449,61 +398,102 @@ public string OutputAssembly
+        ///     parallel{+|-}            - concurrent build
+        ///     warnaserror{+|-}         - report warnings to errors
+        ///     warnaserror{+|-}:strings - report specific warnings to errors
+        ///     warn:number              - warning level (0-4) for CSharp
```
I guess this applies to both csharp and vb. If that's the case, `for CSharp` should be removed.

iSazonov replied:
I don't found the option for VB https://github.com/dotnet/roslyn/blob/master/docs/compilers/Visual%20Basic/CommandLine.md

daxian-dbw replied:
OK, then never mind. #Closed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -568,19 +558,15 @@ private string GetUsingTemplate(Language language)
                 case Language.CSharp:
                     return
-                        "using System;\n" +
-                        "using System.Runtime.InteropServices;\n" +
```
https://github.com/PowerShell/PowerShell/pull/6141/files#r171029964
I want to point it out again that this is a breaking change in that `System` and `System.Runtime.InteropServices` are not included when `-UsingNamespace` is not specified after this change.

iSazonov replied:
If we add this we get errors:
```
PS C:\Users\sie\Documents\GitHub\iSazonov\PowerShell>         $CSharpCode1 = @"
>> using System;
>> using System.Runtime.InteropServices;
>>         namespace Test.AddType
>>         {
>>             public class CSharpTest1$guid
>>             {
>>                 public static int Add1(int a, int b)
>>                 {
>>                     return (a + b);
>>                 }
>>             }
>>         }
>> "@
PS C:\Users\sie\Documents\GitHub\iSazonov\PowerShell> Add-type -TypeDefinition $CSharpCode1
Add-type : (2,1): hidden CS8019: Unnecessary using directive.
using System.Runtime.InteropServices;
^
At line:1 char:1
+ Add-type -TypeDefinition $CSharpCode1
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidData: ((2,1): hidden C...sing directive.:CSDiagnostic) [Add-Type], Exception
+ FullyQualifiedErrorId : SOURCE_CODE_ERROR,Microsoft.PowerShell.Commands.AddTypeCommand

Add-type : (1,1): hidden CS8019: Unnecessary using directive.
using System;
^
At line:1 char:1
+ Add-type -TypeDefinition $CSharpCode1
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidData: ((1,1): hidden C...sing directive.:CSDiagnostic) [Add-Type], Exception
+ FullyQualifiedErrorId : SOURCE_CODE_ERROR,Microsoft.PowerShell.Commands.AddTypeCommand

```

iSazonov replied:
Currently Appveyor CI failed with the same error in log line 3827. 😕 

Can I rebase to fix an error with File.wxs?

daxian-dbw replied:
The error seems caused by a warning that more than one `using system;` is specified. I'm sure there is a way to deal with it.

Feel free to rebase.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -449,61 +398,102 @@ public string OutputAssembly
+        [Parameter(ParameterSetName = FromMemberParameterSetName)]
+        [Parameter(ParameterSetName = FromPathParameterSetName)]
+        [Parameter(ParameterSetName = FromLiteralPathParameterSetName)]
+        public SwitchParameter CompileOnly { get; set; }
```
What if `-CompileOnly` and `-PassThru` are both specified?
As mentioned in some other places, I don't think we need this parameter.

iSazonov replied:
Currently `-PassThru` will be ignored.
The switch came from https://github.com/PowerShell/PowerShell/issues/5725#issuecomment-354164119

daxian-dbw replied:
As I stated in comments above https://github.com/PowerShell/PowerShell/pull/6141#discussion_r173548138, when using `-OutputAssembly`, `Add-Type` only compiles the assembly without loading it. So I don't think `-CompileOnly` is really needed.

iSazonov replied:
Currently we always load [generated assembly](https://github.com/PowerShell/PowerShell/blob/c1c5344a8897262433141ecbc13bb06ac2c4bbef/src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs#L1098). So with `-CompileOnly ` we address new scenario.

iSazonov replied:
Fixed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1090,36 +1259,168 @@ private void CompileSourceToAssembly(string source)
+            {
+                // If the file name does not exist, the source text is set by the user using parameters.
+                // In this case, we assume that the source text is of a small size and we can re-allocate by ToString().
+                hash = st.GetText().ToString().GetHashCode();
```
Based on the [comments](https://github.com/dotnet/roslyn/blob/master/src/Compilers/Core/Portable/Syntax/SyntaxTree.cs#L347) in `SyntaxTree`, you can directly call `ToString()` without having to call `GetText()` first.

iSazonov replied:
Fixed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1090,36 +1259,168 @@ private void CompileSourceToAssembly(string source)
+                foreach (var diagnisticRecord in compilerDiagnostics)
+                {
+                    // We shouldn't specify input and output files in ExtendedOptions parameter
+                    // so suppress errors from Roslyn default command line parcer:
```
Typo --> `parcer`

iSazonov replied:
Fixed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1090,36 +1259,168 @@ private void CompileSourceToAssembly(string source)
+                    //      CS1562: Outputs without source must have the /out option specified
+                    //      CS2008: No inputs specified
+                    //      BC2008: No inputs specified
+                    if (diagnisticRecord.IsSuppressed ||
```
Maybe the check of `diagnisticRecord.IsSuppressed` should be put at the outmost of this method. If `isSuppressed` is true, no need to go through the loop.

iSazonov replied:
Sorry, I don't understand - we get diagnisticRecord in the loop - how we can move it out of the loop?

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +875,381 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+        private void CheckDuplicateTypes(Compilation compilation, out ConcurrentBag<String> newTypes)
+        {
+            AllNamedTypeSymbolsVisitor visitor = new AllNamedTypeSymbolsVisitor(_syntaxTreesHash);
+            visitor.Visit(compilation.Assembly.GlobalNamespace);
```
Seems we have to use the recursion https://github.com/dotnet/roslyn/issues/6138

daxian-dbw replied:
The linked issue is quite old (2015), and it doesn't mention the `IAssemblySymbol.TypeNames` property. If it were me, I will try this property to see if it works.

iSazonov replied:
Tested - `IAssemblySymbol.TypeNames` doesn't return full type names.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +875,381 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+        {
+            foreach (var typeName in newTypes)
+            {
+                s_sourceTypesCache.Add(typeName, _syntaxTreesHash);
```
Yes, I checked this in debuger (but I don't found implementation in Roslyn).

daxian-dbw replied:
Thanks. #Closed

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +875,381 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+            public override void VisitNamespace(INamespaceSymbol symbol)
+            {
+                // Main cycle.
+                symbol.GetMembers().AsParallel().ForAll(s => s.Accept(this));
```
I'd remove the AsParallel at all but but don't found how.

daxian-dbw replied:
If `IAssemblySymbol.TypeNames` works, we won't need the visitor code :)

iSazonov replied:
AsParallel() was replaced with `foreach`.
Fixed.

iSazonov committed:
Fix typos and address some comments

daxian-dbw commented:
@iSazonov I left more comments in many of my existing comments to reply to your responses. I add `#Closed` those of my comments that are closed. For the rest, they are still open and need your response. Thanks!

iSazonov committed:
Use IsWhiteSpace()

iSazonov committed:
Remove errorPosition line

iSazonov committed:
Rename ExtendedOptions  to CompilerOptions

iSazonov committed:
Remove Diagnostics.Assert

iSazonov commented:
@daxian-dbw I'm trying to address your comments but some might miss because Github hides them. 😄  Also the browser freezes on his too large PR.

daxian-dbw commented:
>  some might miss because Github hides them

Every time I revisit this PR, I open every hidden comment left by myself and see if you respond to my last reply. For comments that is closed, a `#closed` tag is added to the last message of the comment thread.

iSazonov commented:
@daxian-dbw I have collected almost all open comments for a quicker response.

1. UsingNamespace in [Parameter(ParameterSetName = "FromSource")] 

I believe the scenario can be important if someone has a portable code template and have to compile in Windows PowerShell and PowerShell Core with different usings. If the scenario is important we should enhance Windows PowerShell too. If no I'll remove this.


2. CompileOnly 

Sorry, I already forget my thoughts - `Add-Type -TypeDefinition $code -OutputAssembly f:\test.dll` in Windows PowerShell loads the dll only if `-PassTru` present. It is hidden logic, this is not an obvious behavior of the -PassTru parameter. If we want load the dll and don't want `-PassTru` we have to use:
```c#
Add-Type -TypeDefinition $code -OutputAssembly f:\test.dll -PassTru > $null
```

So the idea is to add new parameter for better UX.


3. WarnAsError and IgnoreWarnings

In master branch code `-IgnoreWarnings` comply with default behavior and really do nothing.
The master branch code already has a breaking change with 6.0.0 GA (really we haven't any code for `-IgnoreWarnings`) - and the breaking change is that user can not get "warnings as errors" with absent `-IgnoreWarnings` or with `-IgnoreWarnings:$false`.

What is fix we want get now? Should we keep CodeDom (Windows PowerShell) default? It is main question. If so it will be _again a breaking change with 6.0.0 GA_.

Roslyn has [opposite default](https://github.com/dotnet/roslyn/blob/944ae114140ba77cbd4be370bf13a7f758f740b3/src/Compilers/CSharp/Portable/CommandLine/CSharpCommandLineParser.cs#L744).

_I'd prefer use new parameter and new default._

I assumed we could keep the `-IgnoreWarnings` old parameter for backward compatibility, but as I said we've already _full_ broken it and it's better to remove it altogether.


4. Done: ExtendedOptions -> CompilerOptions

5. Removed: `Diagnostics.Assert(defaultRefAssemblies.Count < 150...`

6. `if (!char.IsWhiteSpace(errorLineString[i]))`

I fixed this but ...
This come from original code and I'm afraid the code isn't exactly correct because we're replacing one tab with a single space. Perhaps it was necessary for CodeDom. I hope that Roslyn don't use tabs in error messages. Although these whitespaces can come from the code that it compiles. Therefore we have to clean the error message too or remove this cleanup at all. Thoughts?


7. IAssemblySymbol.TypeNames

Not addressed still

iSazonov committed:
Fix parameters

Remove UsingNamespace from "FromSource" parameter set
Remove CompileOnly parameter
Restore IgnoreWarnings behavior
Fix parameters and tests

iSazonov commented:
CI Appveyor failed on `files.wxs` - I can not fix this.

iSazonov committed:
Fix Files.wxs

iSazonov commented:
@TravisEz13 Thanks for help - Files.wxs was fixed.

iSazonov committed:
Implement unify SourceCodeProcessing method

iSazonov commented:
>IAssemblySymbol.TypeNames

It returns strings with simple type name 😕 - we need full qualify type names.

```c#
$guid = [Guid]::NewGuid().ToString().Replace("-","")

$CSharpCode1 = @"
        namespace Test.AddType1
        {
            public class CSharpTest1$guid
            {
                public static int Add1(int a, int b)
                {
                    return (a + b);
                }
            }
            public class CSharpTest2$guid
            {
                public static int Add2(int a, int b)
                {
                    return (a + b);
                }
            }
        }
        namespace Test.AddType2
        {
            public class CSharpTest1$guid
            {
                public static int Add1(int a, int b)
                {
                    return (a + b);
                }
            }
            public class CSharpTest2$guid
            {
                public static int Add2(int a, int b)
                {
                    return (a + b);
                }
            }
        }
"@

Add-Type -TypeDefinition $CSharpCode1
```
![iassemblysymbol](https://user-images.githubusercontent.com/22290914/37707306-a1f311a2-2cfa-11e8-8767-d222b643d2f9.png)

iSazonov committed:
Replace AsParallel() with foreach

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +875,381 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+            }
+        }
+
+        private void VisualBasicSourceCodeProcessing()
```
Fixed.

iSazonov committed:
Tests -> Pester 4

iSazonov commented:
@daxian-dbw Could you please continue? I hope I addresed most comments. I see only (1) CompileOptions Parameter set, (2) using System.

iSazonov commented:
@daxian-dbw I found that now Roslyn does not ignore unnecessary directives as it was in previuos compiler versions, but uses another API for code analyzers to exclude generated files with unused directives.
https://github.com/dotnet/roslyn/issues/5162

We can do:
```powershell
-CompilerOptions "/nowarn:CS8019"
```
but should we do this by default?

I believe we shouldn't undoubtedly add `using System;` in Add-Type generated code as requested above.

iSazonov committed:
Some fixes

Fix UsingNamespace after excluding from FromSourceParameterSet.
IgnoreWarnings override CompilerOptions
Report errors before assembly loading
Skip diagnostic records wth severity equal to Hidden (like CS8019
UnusedUsingDirective)

iSazonov committed:
Suppress unneeded output in test

iSazonov commented:
@daxian-dbw Seems I addressed all comments except one.

I don't understand that is breaking change if we don't add `using System;` in Add-Type generated code as requested above. Could you give a test sample?

iSazonov committed:
Fix usings

iSazonov commented:
@daxian-dbw Seems I addressed all comments.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +853,370 @@ private Assembly LoadAssemblyHelper(string assemblyName)
                     {
+                        // TODO:  We could use Assembly.LoadFromStream() in future.
+                        // See https://github.com/dotnet/corefx/issues/26994
                         ms.Flush();
```
`Flush()` on `MemoryStream` does nothing, see: http://source.dot.net/#System.Private.CoreLib/shared/System/IO/MemoryStream.cs,179

iSazonov replied:
I think we should not rely on the knowledge of the current implementation - what if it changes tomorrow?
We'll remove this in Assembly.LoadFromStream() time.

daxian-dbw replied:
As long as you are using a memory stream, `flush` won't do anything. It's by design because there is no place you need to flush the content to.

iSazonov replied:
Removed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1090,36 +1226,172 @@ private void CompileSourceToAssembly(string source)
+                        CacheNewTypes(newTypes);
+                        CacheAssemply(assembly);
+
+                        if (PassThru)
```
Don't need to check `PassThru` again.

iSazonov replied:
Fixed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1090,36 +1226,172 @@ private void CompileSourceToAssembly(string source)
+                        continue;
+                    }
+
+                    IsError |= diagnisticRecord.Severity == DiagnosticSeverity.Error || (diagnisticRecord.IsWarningAsError && diagnisticRecord.Severity == DiagnosticSeverity.Warning);
```
How about changing it to:
```
if (!IsError)
{
    IsError = diagnisticRecord.Severity == DiagnosticSeverity.Error || 
              (diagnisticRecord.IsWarningAsError && diagnisticRecord.Severity == DiagnosticSeverity.Warning);
}
```

iSazonov replied:
Fixed.

iSazonov commented:
@TravisEz13 
>Do we have enough ask to add the 5MB for VB support?

My motivation was to achieve compatibility with Windows PowerShell. 
Also we have requestes to add JS and F#.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -431,79 +379,111 @@ public string OutputAssembly
+        [Parameter(ParameterSetName = FromMemberParameterSetName)]
+        [Parameter(ParameterSetName = FromPathParameterSetName)]
+        [Parameter(ParameterSetName = FromLiteralPathParameterSetName)]
+        [Alias("IW")]
```
I don't like aliases at all.
Is the removing a breaking change?

daxian-dbw replied:
The `IW` currently doesn't exist, so let's not add it here.
For other existing aliases, it will be a breaking change to remove, so let's keep them.

iSazonov replied:
Alias removed.

iSazonov committed:
Fix PassThru

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -431,79 +379,111 @@ public string OutputAssembly
+        [Parameter(ParameterSetName = FromMemberParameterSetName)]
+        [Parameter(ParameterSetName = FromPathParameterSetName)]
+        [Parameter(ParameterSetName = FromLiteralPathParameterSetName)]
+        [Parameter(ParameterSetName = FromAssemblyNameParameterSetName)]
```
Fixed.

iSazonov committed:
Add [ValidateNotNullOrEmpty] in CompilerOptions

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -431,79 +379,111 @@ public string OutputAssembly
+        [Parameter(ParameterSetName = FromMemberParameterSetName)]
+        [Parameter(ParameterSetName = FromPathParameterSetName)]
+        [Parameter(ParameterSetName = FromLiteralPathParameterSetName)]
+        public string[] CompilerOptions { get; set; }
```
Fixed.

iSazonov committed:
Use nameof

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -714,10 +606,16 @@ protected override void EndProcessing()
             // Generate an error if they've specified an output
             // assembly type without an output assembly
-            if (String.IsNullOrEmpty(outputAssembly) && outputTypeSpecified)
+            if (String.IsNullOrEmpty(_outputAssembly) && this.MyInvocation.BoundParameters.ContainsKey("OutputType"))
```
Fixed.
Opened #6548 for cleanups.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -726,63 +624,65 @@ protected override void EndProcessing()
+                // File extension is ".DLL" (ParameterSetName = FromPathParameterSetName or FromLiteralPathParameterSetName).
+                LoadAssemblies(_paths);
+            }
+            else if (AssemblyName != null && AssemblyName.Length > 0)
```
Actually we check a parameter set name.
I'd use `this.MyInvocation.BoundParameters.ContainsKey(nameof(AssemblyName))` or `ParameterSetName == FromAssemblyNameParameterSetName`.

daxian-dbw replied:
Yes, checking the parameter set is fine.
Make sure you remove the comment `// ParameterSetName = FromAssemblyNameParameterSetName.` after you change to checking parameter set.

iSazonov committed:
Remove extra spaces

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -726,63 +624,65 @@ protected override void EndProcessing()
+        private  const string FromSourceParameterSetName = "FromSource";
+        private  const string FromPathParameterSetName = "FromPath";
+        private  const string FromLiteralPathParameterSetName = "FromLiteralPath";
+        private  const string FromAssemblyNameParameterSetName = "FromAssemblyName";
```
Fixed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +853,370 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+            return null;
+        }
+
+        private SyntaxTree ParseText(SourceText sourceText, ParseOptions parseOptions, String path = "")
```
Fixed.

iSazonov committed:
Fix case

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +853,370 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+            {
+                case Language.CSharp:
+                    CSharpCompilationOptions csharpCompilationOptions;
+                    csharpCompilationOptions = new CSharpCompilationOptions(OutputAssemblyTypeToOutputKind(OutputType));
```
Fixed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +853,370 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+
+                case Language.VisualBasic:
+                    VisualBasicCompilationOptions vbCompilationOptions;
+                    vbCompilationOptions = new VisualBasicCompilationOptions(outputKind: OutputAssemblyTypeToOutputKind(OutputType));
```
Fixed.

iSazonov committed:
Use var

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +853,370 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+
+                parseOptions = arguments.ParseOptions;
+
+                compilationOptions = arguments.CompilationOptions.WithOutputKind(OutputAssemblyTypeToOutputKind(OutputType));
```
> Defining a precedence order is confusing.

We could document this and keep the flexibility. Also I am thinking about `Add-Type -CompilerOption $defaultOptions` - we could overlap _default options_ with explicit parameters.

daxian-dbw replied:
Then please update the PR summary to emphasize that `-CompilerOptions` can be specified along with other options like `-OutputAssembly`, `-OutputType` and `-IgnoreWarnings`, and the explicit setting parameters will take precedence over the same setting in `-CompilerOptions`.

iSazonov replied:
PR description was updated.

iSazonov committed:
Remove extra lines

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +853,370 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+            EmitOptions emitOptions = null;
+
+            if (CompilerOptions != null)
+            {
```
Broad soul :smile:
Fixed.

iSazonov committed:
Remove code duplication

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +853,370 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+            if (!IgnoreWarnings.IsPresent)
+            {
+                compilationOptions = compilationOptions.WithGeneralDiagnosticOption(defaultDiagnosticOption);
+            }
```
Removed from `GetDefaultCompilationOptions`.

iSazonov committed:
Move WriteVerbose() on better position

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +853,370 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+                        WriteTypes(assembly);
+                    }
+
+                    WriteVerbose(AddTypeStrings.AlreadyCompiledandLoaded);
```
Fixed.

iSazonov committed:
Remove PassThru check

iSazonov committed:
Use if

iSazonov committed:
Merge branch 'master' into addtype-enhance1

iSazonov committed:
Fix PassThru

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -317,48 +274,39 @@ private void ProcessPaths(List<string> resolvedPaths)
+        [Parameter(ParameterSetName = FromSourceParameterSetName)]
+        [Parameter(ParameterSetName = FromMemberParameterSetName)]
+        [Parameter(ParameterSetName = FromPathParameterSetName)]
+        [Parameter(ParameterSetName = FromLiteralPathParameterSetName)]
```
Then what if you have VB in the `.tmp` file and specify `-Language CSharp` :)?
Plus, `Add-Type` will throw an error for any extension name other than `.cs`, `.vb` or `.dll` at ProcessPaths(List<string> resolvedPaths).

iSazonov replied:
In my scenario customer can rename .tmp -> .cs and resolve the issue. So I addressed your request.
Fixed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +850,356 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+                case Language.CSharp:
+                    var csharpCompilationOptions = new CSharpCompilationOptions(OutputAssemblyTypeToOutputKind(OutputType));
+
+                    return csharpCompilationOptions;
```
Then, why not just `return new CSharpCompilationOptions(OutputAssemblyTypeToOutputKind(OutputType));`?

iSazonov replied:
Fixed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +850,356 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+                case Language.VisualBasic:
+                    var vbCompilationOptions = new VisualBasicCompilationOptions(outputKind: OutputAssemblyTypeToOutputKind(OutputType));
+
+                    return vbCompilationOptions;
```
Same here, maybe just `return new VisualBasicCompilationOptions(outputKind: OutputAssemblyTypeToOutputKind(OutputType));`?

iSazonov replied:
Fixed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1090,36 +1209,173 @@ private void CompileSourceToAssembly(string source)
+
                 if (emitResult.Success)
                 {
+                    Assembly assembly = Assembly.LoadFrom(_outputAssembly);
```
You cannot load the assembly unconditionally. Only load the assembly when `-PassThru` is specifed.

iSazonov replied:
Fixed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +850,356 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+            }
+        }
+
+        private CommandLineArguments DefaultArgsParse(IEnumerable<string> args)
```
The method name `DefaultArgsParse` should be renamed to `ParseCompilerOption`.

iSazonov replied:
Renamed.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +850,356 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+                HandleCompilerErrors(arguments.Errors);
+
+                parseOptions = arguments.ParseOptions;
+                compilationOptions = arguments.CompilationOptions.WithOutputKind(OutputAssemblyTypeToOutputKind(OutputType));
```
What if `-OutputType` is not specified? Shall we respect the `-CompilerOption` or the default value of `-OutputType`?

iSazonov replied:
Roslyn [default](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/CommandLine/CSharpCommandLineParser.cs,79) is `OutputKind.ConsoleApplication`. But we should have `Library` default. We don't custom parsing so we have to ignore `CompilerOptions` and always use `OutputType`.

PR description was updated.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +850,356 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+            return null;
+        }
+
+        private CompilationOptions GetDefaultCompilationOptions()
```
The name should be renamed to `GetDefaultCompilerOption` since the option is used by parser, compilation, and emition.

iSazonov replied:
But it returns `CompilationOptions` not `CompilerOptions`.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/AddType.cs:
```diff
@@ -1010,77 +850,356 @@ private Assembly LoadAssemblyHelper(string assemblyName)
+            return null;
+        }
+
+        private SyntaxTree ParseText(SourceText sourceText, ParseOptions parseOptions, string path = "")
```
nit: Maybe rename this method to `ParseSourceText`? It's more readable.

iSazonov replied:
Renamed.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Add-Type.Tests.ps1:
```diff
@@ -28,41 +28,242 @@ Describe "Add-Type" -Tags "CI" {
+
+        $cmdlet = "Get-CompileThing$guid"
+
+        Add-Type -TypeDefinition $code -OutputAssembly $outFile | Should -BeNullOrEmpty
```
Could you please add a test to verify that when `-OutputAssembly` is specified, the generated assembly is not loaded by default, unless `-PassThru` is specified too?

iSazonov replied:
1. Added
2. Please see Line 153.

iSazonov replied:
1. Added
2. Please see Line 153.

iSazonov replied:
1. Added
2. Please see Line 153.

iSazonov committed:
Remove [Alias("IW")]

iSazonov committed:
Use ParameterSetName == FromAssemblyNameParameterSetName

iSazonov committed:
Remove -Language from path parameter sets

iSazonov committed:
Remove ms.Flush()

iSazonov committed:
Fix return

iSazonov committed:
Rename DefaultArgsParse to ParseCompilerOption

iSazonov committed:
Rename ParseText to ParseSourceText

iSazonov committed:
Fix PassThru

iSazonov committed:
Fix typo

iSazonov committed:
Fix test

iSazonov committed:
Fix typo in merge conflict resolve

daxian-dbw commented:
@adityapatwardhan and @TravisEz13 Could you please quickly review the tests? Not sure if the `Add-Type` tests should be made CI ...

@TravisEz13 and @vors brought up the concern about adding back `VB` support (with 5MB new assembly). So add `Committee-Review` label to get a conclusion from the committee.

@iSazonov Can you please fix the failed Travis CI build?

joeyaiello commented:
We just had a long conversation within the @PowerShell/powershell-committee about this. Some high-level thoughts:
* First, thank you for doing this work here. This is clearly a huge PR, and we know how much of a pain that can be.
* We have no clue about the magnitude of demand for VB in Add-Type. @iSazonov: are you aware of customer use cases or do you personally have a use case for VB that you can elaborate on here?
* We're a *little* worried about adding 5 MBs to the size footprint. Today, $PSHome is about 125 MBs, and we have a strong desire to shrink that over time. 5 MBs might be a LOT for little benefit. 
* Given that this PR has been open for a while, I think it makes a lot of sense to talk to customers at the PowerShell Summit NA and PSconfEU over the next two weeks to get an understanding of the demand for VB support. 

In the event that we decide not to accept VB for 6.1, @daxian-dbw has said that he could take on the work of removing the VB code. But in the meantime, we'd like to wait until after the two conferences to merge this PR.

iSazonov commented:
@joeyaiello 
1. My main motivation is to achieve maximum compatibility with Windows PowerShell Add-Type. Released version has too many incompatibilities (even without considering VB).

2. I don't know about customer cases. I would addressed this question to MSFT and above all to _VB team_. This language has very long history. If MSFT wants to use it further _widely_, it should be here.
As a system administrator, I can find _many implementations_ of algorithms on VB and easily inject them into a script without having to rewrite to C#. This is especially important if the algorithm is very complex.

If someone is developing a large project on VB and uses scripts for automation (as we do in this project), then I guess they will likely use VB in scripts too.  

Also we have requests (in the repo Issues) to add F# and JS support.

3. 5 MB. Most of the size PowerShell is .Net Core. Each new version will have a larger size. I don't think we can significantly reduce the number of dependencies. We could only change the model with Self-contained on Framework-dependent to get small distributive.

daxian-dbw commented:
I have to delete 30 old comments in order to make the web browser able to open this PR page. Wait for the @PowerShell/powershell-committee's final call to merge this PR.

SteveL-MSFT commented:
@daxian-dbw this is a long PR, can you summarize what the committee needs to review?

daxian-dbw commented:
@SteveL-MSFT I need the committee's approval to merge the PR as is (with the VB related code).
If the committee decides to not have VB support in `Add-Type`, then we can use a separate PR to clean it up.

SteveL-MSFT commented:
@PowerShell/powershell-committee is ok with taking this PR as-is and having a separate discussion about removing VB support

daxian-dbw commented:
@iSazonov Thanks for the hard work to get `Add-Type` refactored!

iSazonov commented:
@daxian-dbw Thanks for great comments and help!

daxian-dbw commented:
Remove `Breaking-Change` label because no breaking change is introduced in this PR.

daxian-dbw commented:
Add back the `Breaking-Change` label as @dantraMSFT pointed out that some error strings were removed meaning that we are using different `FullyQualifiedName` for those error conditions.

rjmholt commented:
🔨

iSazonov commented:
We have never considered changing FullyQualifiedName as breaking change. However, I agree to leave this label because updates to the cmdlet are significant.
